lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2084,14 +2084,20 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
     checkGenericParamAccess(TC, CD->getGenericParams(), CD);
 
     if (CD->hasSuperclass()) {
-      Type superclass = CD->getSuperclass();
+      const NominalTypeDecl *superclassDecl =
+          CD->getSuperclass()->getAnyNominal();
+      // Be slightly defensive here in the presence of badly-ordered
+      // inheritance clauses.
       auto superclassLocIter = std::find_if(CD->getInherited().begin(),
                                             CD->getInherited().end(),
                                             [&](TypeLoc inherited) {
         if (!inherited.wasValidated())
           return false;
-        return inherited.getType().getPointer() == superclass.getPointer();
+        return inherited.getType()->getAnyNominal() == superclassDecl;
       });
+      // Sanity check: we couldn't find the superclass for whatever reason
```

Comment:
What about this?

```
class Base {}
protocol P {}
typealias T = Base & P
class Derived : T {}
```

Reply:
I would just remove the check that the superclass appears in the inheritance clause and see if anything breaks

Reply:
The problem is we want a TypeRepr if at all possible. In your example (which I should add) we can warn on `T` because we're checking the sugared type, but we *shouldn't* warn on some protocol that isn't the superclass.

Maybe I should just take the first type that has a class in it.

Reply:
I see. You can check `isExistentialType()`, and if so, look at `getExistentialLayout().superclass`. 

However I realized in the typealias case, we should be preventing you from subclassing an internal typealias if the class is public (can you test that?)

Similarly, a public typealias is OK to subclass, and it cannot reference an internal class anyway, because we would have diagnosed when the typealias was declared.